lib/repl.js

```diff
@@ -1442,6 +1445,11 @@ function complete(line, callback) {
       const bits = StringPrototypeSplit(match, '.');
       filter = ArrayPrototypePop(bits);
       expr = ArrayPrototypeJoin(bits, '.');
+      // If filter starts with a lower case char we will ignore case when
+      // filtering completion options
+      useCaseSensitiveFilterForCompletionGroups =
+        StringPrototypeToLocaleLowerCase(StringPrototypeCharAt(filter, 0)) !==
+        StringPrototypeCharAt(filter, 0);
```

Comment:
Did you consider using "contains only lower case chars" as the condition instead (similar to VIM's `smartcase` option)?

Reply:
I haven't, I never used VIM extensively so was not aware of how it handles autocomplete, TIL! I do explain in the linked issue https://github.com/nodejs/node/issues/41631 why I implemented it that way and what are the alternatives that I considered, but I'm totally open to changing the behavior here to whatever Node.js team would consider the best approach

Reply:
It's just that having to start with an uppercase letter is a pretty big constraint for enabling case sensitivity, knowing that most of JS code out there is written using camelCase.

Reply:
This should work I think:
```suggestion
      // If `filter` contains no upper case char, we ignore case when
      // filtering completion options.
      useCaseSensitiveFilterForCompletionGroups =
        RegExpPrototypeExec(/[A-Z]/, filter) !== null;
```

Reply:
@aduh95 sorry, struggling to catch up with all the notifications recently, thanks for the ping!

@targos I totally get the camelCase argument, at the same time all browsers follow this behavior and all browser devtools combined are probably the most used JavaScript REPLs. FWIW the main downside to this autocomplete behavior I see is more than one autocomplete options for something like a `function` keyword, but as it's all lowercase, making this check stricter per your suggestion will not cover this case as `function` is all lowercase. As this got a few approvals already is it a good enough indicator that currently implemented approach will work for Node.js REPL in your opinion?

Also [as I mentioned in the issue](https://github.com/nodejs/node/issues/41631#issuecomment-1019286874) I think the "too many options" issue is not really an issue here, what's inconvenient is that when you get multiple options you can't really select from them, I would be eager to try implementing selection logic next as a further improvement of the REPL autocomplete UX

Reply:
> all browsers follow this behavior

All browsers have different behavior AFAICT:

- Chromium: always case insensitive.
- Safari: always case sensitive.
- Firefox: case sensitive only if the first letter is upper case.

To test that, I've declared to following object:

```js
const d = { Toe: 'toa', toString: 'str', tooling: 'test', tostringify: 'test' }
```

And tried to type `d.t`, `d.T`, `d.to`, `d.To`, `d.TO`, `d.tos`, `d.Tos`, `d.toS`, `d.ToS`, and `d.TOS`. Are you seeing different results?

Firefox being arguably the least intuitive of all three imo: either we want case sensibility, either we don't want it, it feels weird to use the first char to discriminate. That being said, I don't personally have any preference over one or the other. I agree it makes sense not to follow Vim's behavior.

Reply:
Huh, yeah, you're right, I think I got confused when testing it in different browsers.

Reply:
@aduh95 updated the code to make autocomplete always case-insensitive, does this looks good to you now?